### PR TITLE
Allow for CIP8 that bypasses CIP3

### LINKED ISSUE #7848: CIP3 Deprecation
This is required to ensure an account can be storage sources from multiple writers. The existing structure of nesting CIP8 storage urls (secondary file) within CIP3 metadata storage endpoint (primary file) means that a new writer would need to copy, update to add additional secondary file, and re-host the primary file themselves. 

To reduce adoption complexity, we should just have the array/dictionary of storage endpoints in the contract itself.

Options:
- Migrate to storage array in Account profile (may require iterating through multiple storage nodes for needed info)
- Dictionary mapping for storage on Accounts (could be mapped by "client"/"writer")

https://www.notion.so/clabsco/Deprecating-CIP3-5b85e8025c8a4aefa86814f8dcf36825

### PR DESCRIPTION
**Pull request by m-chrzan**
### Description

CIP8 (offchain storage) is currently implemented on top of CIP3 (offchain metadata). DApp development can be easier if additional CIP8 storage roots can be added via an onchain transaction rather than modifying the CIP3 record.

This PR introduces a new storage variable on the Accounts smart contract that can be used to bypass CIP3 for registering new CIP8 storage roots.

### Tested

New unit tests.

### Related issues

- Fixes #7848

### Backwards compatibility

Adds new storage variable and related methods, doesn't currently affect anything else. This would be a first step in deprecating CIP3.

### CODE REVIEW

File: packages/protocol/contracts/common/Accounts.sol [hunk 1]
```diff
@@ -238,6 +243,66 @@ contract Accounts is
...
+   * multiple storage roots at a time.
+   */
+  function removeStorageRoot(uint256 index) external {
+    require(isAccount(msg.sender), "Unknown account");
+    require(index < offchainStorageRoots[msg.sender].length);
+    uint256 lastIndex = offchainStorageRoots[msg.sender].length - 1;
+    bytes memory url = offchainStorageRoots[msg.sender][index];
+    offchainStorageRoots[msg.sender][index] = offchainStorageRoots[msg.sender][lastIndex];
+    offchainStorageRoots[msg.sender].length--;
+    emit OffchainStorageRootRemoved(msg.sender, url);
```
Comment (yorhodes) [inline packages/protocol/contracts/common/Accounts.sol]: I think index should also be emitted here for cache invalidation

File: packages/protocol/contracts/common/Accounts.sol [hunk 2]
```diff
@@ -238,6 +243,66 @@ contract Accounts is
...
+    require(isAccount(account), "Unknown account");
+    uint256 numberRoots = offchainStorageRoots[account].length;
+    uint256 totalLength = 0;
+    for (uint256 i = 0; i < numberRoots; i++) {
+      totalLength += offchainStorageRoots[account][i].length;
+    }
+
+    bytes memory concatenated = new bytes(totalLength);
+    uint256 lastIndex = 0;
+    uint256[] memory lengths = new uint256[](numberRoots);
```
Comment (yorhodes) [inline packages/protocol/contracts/common/Accounts.sol:293]: does storing the roots concatenated make more sense?

File: packages/protocol/contracts/common/Accounts.sol [hunk 3]
```diff
@@ -857,6 +922,56 @@ contract Accounts is
...
+   * @notice Getter for the storage roots of multiple accounts.
+   * @param accountsToQuery The addresses of the accounts to get the storage roots for.
+   * @return (
+   *          data - all strings concatenated
+   *          rootLengths[] - the length of each string in `data` in bytes
+   *          accountRootNumbers[] - the number of storage roots per account
+   *         )
+   */
+
+  function batchGetOffchainStorageRoots(address[] calldata accountsToQuery)
```
Comment (yorhodes) [inline packages/protocol/contracts/common/Accounts.sol]: maybe lacking context but don't see why this is useful
I suppose to reduce number of round trips but the RPC supports batching calls too

File: packages/protocol/specs/accounts.spec [hunk 4]
```diff
@@ -205,6 +205,11 @@ rule viewFunctionsDoNotRevert(method f) filtered { f ->
 	// some functions we ignore, and the reasons:
 	&& f.selector != hasAuthorizedSigner(address,string).selector // Calldatasize may not match
 	&& f.selector != batchGetMetadataURL(address[]).selector // Calldatasize may not match
+
+  // These require an account to exist
```
Comment (yorhodes) [inline packages/protocol/specs/accounts.spec:209]: 🙏

**Review by yorhodes — APPROVED** [review]
few questions

File: packages/protocol/contracts/common/Accounts.sol (hunk 2 above)
Comment (m-chrzan) [reply packages/protocol/contracts/common/Accounts.sol:293]: I believe that would make adding/removing roots more complicated?

File: packages/protocol/contracts/common/Accounts.sol (hunk 3 above)
Comment (m-chrzan) [reply packages/protocol/contracts/common/Accounts.sol]: I believe avoiding round trips was the main goal with this. If this is supported out-of-contract that would be ideal (especially given how complicated this function ended up being), let's discuss this with the wallet team that would be consuming this.

Comment (isabellewei) [reply packages/protocol/contracts/common/Accounts.sol]: agreed that I can't really envision a specific use case for this, we can always batch from the wallet side

**Review by isabellewei — COMMENTED** [review]
This makes sense to consume from a wallet perspective, thanks! Could we also add a function in `contractkit` to return the roots as an array :)

**Comment by m-chrzan** [comment]
> Could we also add a function in `contractkit` to return the roots as an array :)

Let's merge this PR without increasing scope. Created #8850.

**Pull request merged**